packages/shared-components/src/components/onto-repository-selector/onto-repository-selector.tsx

```diff
@@ -0,0 +1,102 @@
+import {Component, Host, h, State} from '@stencil/core';
+import {ServiceProvider, RepositoryService, RepositoriesList} from "@ontotext/workbench-api";
+import {DropdownItem} from '../../models/dropdown/dropdown-item';
+import {DropdownItemAlignment} from '../../models/dropdown/dropdown-item-alignment';
+import {Subscription} from 'rxjs';
+import {Repository} from '@ontotext/workbench-api';
+
+@Component({
+  tag: 'onto-repository-selector',
+  styleUrl: 'onto-repository-selector.scss',
+  shadow: false,
+})
+/**
+ * A StencilJS component that renders a repository selector dropdown. This component fetches the list of repositories
+ * from the `RepositoryService` and displays them in a dropdown. It allows users to select a repository, and automatically
+ * updates the current repository when a selection is made.
+ *
+ * @example
+ * <onto-repository-selector></onto-repository-selector>
+ */
+export class OntoRepositorySelector {
+  private repositoryService: RepositoryService;
+  private repositories: RepositoriesList;
+  private readonly subscriptions: Subscription[] = [];
+
+  /**
+   * A list of dropdown items representing repositories, updated when repositories change.
+   */
+  @State() items: DropdownItem<Repository>[] = [];
+
+  /**
+   * The currently selected repository, which will be shown in the dropdown button.
+   */
+  @State() currentRepository: Repository;
+
+  constructor() {
+    this.repositoryService = ServiceProvider.get(RepositoryService);
+    this.repositoryService.loadRepositories();
+    this.subscriptions.push(this.subscribeToRepositoriesChanged());
+    this.subscriptions.push(this.subscribeToSelectedRepositoryChanged());
+  }
+
+  /**
+   * Cleans up the subscriptions when the component is removed from the DOM to avoid memory leaks.
+   * This method unsubscribes from both the repository change and selected repository change events.
+   */
+  disconnectedCallback(): void {
+    this.subscriptions.forEach((subscription) => subscription.unsubscribe());
+  }
+
+  render() {
+    return (
+      <Host>
+        <onto-dropdown
+          class='onto-repository-selector'
+          onValueChanged={(newRepository: any) => this.onRepositoryChanged(newRepository)}
+          dropdownButtonName={<onto-repository-selector-button repository={this.currentRepository}></onto-repository-selector-button>}
+          dropdownAlignment={DropdownItemAlignment.RIGHT}
+          items={this.items}>
+        </onto-dropdown>
+      </Host>
+    );
+  }
+
+  private subscribeToRepositoriesChanged(): Subscription {
+    return this.repositoryService.onRepositoriesChanged()
+        .subscribe((repositories: RepositoriesList) => {
+          this.repositories = repositories || new RepositoriesList();
```

Comment:
Why the service doesn't guarantee that the value is always a RepositoryList instead of spreading such defaults around?

Reply:
done